pkg/metrics/collector.go

```diff
@@ -96,21 +102,266 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
+		if len(fields) < 11 {
+			return nil, fmt.Errorf("unexpected format in stat file")
+		}
+
+		return parseDriveStats(fields)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("reading stat file: %w", err)
+	}
+
+	return nil, fmt.Errorf("stat file is empty")
+}
+
+func parseDriveStats(fields []string) (*driveStats, error) {
+	var stats driveStats
+	var err error
+
+	stats.ReadSectors, err = strconv.ParseInt(fields[2], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing ReadSectors: %w", err)
+	}
+	stats.ReadTicks, err = strconv.ParseInt(fields[3], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing ReadTicks: %w", err)
+	}
+	stats.WriteSectors, err = strconv.ParseInt(fields[6], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing WriteSectors: %w", err)
+	}
+	stats.WriteTicks, err = strconv.ParseInt(fields[7], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing WriteTicks: %w", err)
+	}
+	stats.TimeInQueue, err = strconv.ParseInt(fields[9], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing TimeInQueue: %w", err)
+	}
+
+	return &stats, nil
+}
+
+func getSectorSize(device string) (int64, error) {
+	// Construct the file path to the sector size file
+	filePath := "/sys/block/" + device + "/queue/hw_sector_size"
+
+	// Open the file
+	file, err := os.Open(filePath)
+	if err != nil {
+		// Return the error and a default sector size
+		return 512, fmt.Errorf("error opening sector size file: %v", err)
+	}
+	defer file.Close()
+
+	// Read the sector size
+	scanner := bufio.NewScanner(file)
+	if scanner.Scan() {
+		sectorSizeStr := scanner.Text()
+		sectorSize, err := strconv.ParseInt(sectorSizeStr, 10, 64)
+		if err != nil {
+			// Return the error and a default sector size
+			return 512, fmt.Errorf("error parsing sector size: %v", err)
+		}
+		return sectorSize, nil
+	}
+
+	if err := scanner.Err(); err != nil {
+		// Return the error and a default sector size
+		return 512, fmt.Errorf("error reading sector size: %v", err)
+	}
+
+	return 512, nil // Default sector size
+}
+
 // Collect is called by Prometheus registry when collecting metrics.
 func (c *metricsCollector) Collect(ch chan<- prometheus.Metric) {
 	ctx, cancelFunc := context.WithCancel(context.Background())
 	defer cancelFunc()
 
-	resultCh := volume.NewLister().
+	// Collecting volume statistics
+	volumeResultCh := volume.NewLister().
 		NodeSelector([]directpvtypes.LabelValue{directpvtypes.ToLabelValue(string(c.nodeID))}).
 		List(ctx)
-	for result := range resultCh {
+	for result := range volumeResultCh {
 		if result.Err != nil {
-			return
+			continue
 		}
 
 		if result.Volume.Status.TargetPath != "" {
 			c.publishVolumeStats(ctx, &result.Volume, ch)
 		}
 	}
+
+	// Collecting drive statistics
+	driveResultCh := drive.NewLister().
+		NodeSelector([]directpvtypes.LabelValue{directpvtypes.ToLabelValue(string(c.nodeID))}).
+		List(ctx)
+	for result := range driveResultCh {
+		if result.Err != nil {
+			continue
```

Comment:
Why `continue`? do you mean `break`?

Reply:
I used `continue` to ensure we process all drives, even if an error occurs on one of them.

Reply:
Once the error is thrown, we cannot get values further. `break` is logically correct thing to do.